src/main/cpp/jni/ir_sahab_regexmatcher_RegexMatcher.cc

```diff
@@ -22,71 +24,87 @@ const char* java_pattern_preparation_exception_path = "ir/sahab/regexmatcher/exc
-        JNIEnv* jenv, jobject jobj, jlong jinstance_handle) {
-    instances.erase(static_cast<unsigned int>(jinstance_handle));
+        JNIEnv* jenv, jobject jobj, jlong jinstance_id) {
+    instances_map_mutex.lock();
+    instances.erase(static_cast<unsigned int>(jinstance_id));
+    instances_map_mutex.unlock();
 }
 
 JNIEXPORT void JNICALL Java_ir_sahab_regexmatcher_RegexMatcher_addPattern(
-        JNIEnv* jenv, jobject jobj, jlong jinstance_handle, jlong jpattern_id, jstring jpattern,
+        JNIEnv* jenv, jobject jobj, jlong jinstance_id, jlong jpattern_id, jstring jpattern,
         jboolean is_case_sensitive) {
-    HyperscanWrapper* instance = GetHyperscanInstance(jenv, jinstance_handle);
-    if (instance == nullptr)
-        return;
+    HyperscanWrapper* instance = GetHyperscanInstance(jenv, jinstance_id);
+    if (instance == nullptr) {
+        return; // C++ continues to work after Java exception is thrown
+    }
 
     auto pattern = jenv->GetStringUTFChars(jpattern, nullptr);
-    instance->AddPattern(static_cast<unsigned int>(jpattern_id), pattern, is_case_sensitive);
+    if (pattern == nullptr) {
+        ThrowJavaException(jenv, java_assertion_error_path, "Unable to convert java 'pattern' string to cpp string!");
+        return; // C++ continues to work after Java exception is thrown
+    }
+    instance->AddPattern(static_cast<unsigned int>(jpattern_id), pattern, (bool)(is_case_sensitive == JNI_TRUE));
     jenv->ReleaseStringUTFChars(jpattern, pattern);
 }
 
 JNIEXPORT jboolean JNICALL Java_ir_sahab_regexmatcher_RegexMatcher_removePattern(
-        JNIEnv* jenv, jobject jobj, jlong jinstance_handle, jlong jpattern_id) {
-    HyperscanWrapper* instance = GetHyperscanInstance(jenv, jinstance_handle);
-    if (instance == nullptr)
-        return JNI_FALSE;
+        JNIEnv* jenv, jobject jobj, jlong jinstance_id, jlong jpattern_id) {
+    HyperscanWrapper* instance = GetHyperscanInstance(jenv, jinstance_id);
+    if (instance == nullptr) {
+        return JNI_FALSE; // C++ continues to work after Java exception is thrown
+    }
 
-    return instance->RemovePattern(static_cast<unsigned int>(jpattern_id)) ?
-            JNI_TRUE : JNI_FALSE;
+    return instance->RemovePattern(static_cast<unsigned int>(jpattern_id)) ? JNI_TRUE : JNI_FALSE;
 }
 
 JNIEXPORT void JNICALL Java_ir_sahab_regexmatcher_RegexMatcher_preparePatterns(
-        JNIEnv* jenv, jobject jobj, jlong jinstance_handle) {
-    HyperscanWrapper* instance = GetHyperscanInstance(jenv, jinstance_handle);
-    if (instance == nullptr)
-        return;
+        JNIEnv* jenv, jobject jobj, jlong jinstance_id) {
+    HyperscanWrapper* instance = GetHyperscanInstance(jenv, jinstance_id);
+    if (instance == nullptr) {
+        return; // C++ continues to work after Java exception is thrown
+    }
 
-    auto result = static_cast<jlong>(instance->CompilePatterns());
+    auto result = instance->CompilePatterns();
```

Comment:
CompilePattern should return int64_t instead of int because it will return id of problematic pattern which is long!